### PR TITLE
fix(server): small string allocations only under 256 bytes str

### PR DESCRIPTION
**Pull request by adiholden**
When allocating small string if SegmentAllocator::address_table_ is bigger than 2^12 we will get wrong pointer when translating the address SegmentAllocator::Translate
This led to crash in dragonfly in DbSlice::AddExpire
CHECK(db_arr_[db_ind]->expire.Insert(main_it->first.AsRef(), ExpirePeriod(delta)).second);
As the key was not in the dash table but because the key string that was fetched from main_it->first.AsRef() was not the correct key that was in the insert flow as the address table was too big

### CODE REVIEW

File: src/core/small_string.h [hunk 1]
```diff
@@ -55,7 +57,7 @@ class SmallString {
   char prefix_[kPrefLen];
 
   uint32_t small_ptr_;  // 32GB capacity because we ignore 3 lsb bits (i.e. x8).
```
Comment (adiholden) [inline src/core/small_string.h:59]: @romange still did not understand fully how this should be changed :(

File: src/core/compact_object.cc [hunk 2]
```diff
@@ -790,8 +790,8 @@ void CompactObj::SetString(std::string_view str) {
     }
   }
 
-  if (kUseSmallStrings) {
-    if ((taglen_ == 0 && encoded.size() < (1 << 13))) {
+  if (kUseSmallStrings && SmallString::CanAllocate()) {
```
Comment (romange) [inline src/core/compact_object.cc:793]: I suggest reorder CanAllocate after comparing the  SmallString::kMaxSize. Or, alternatively, extend `CanAllocate` to pass the length parameter and check it there first

File: src/core/compact_object.cc [hunk 3]
```diff
@@ -790,8 +790,8 @@ void CompactObj::SetString(std::string_view str) {
...
+    if ((taglen_ == 0 && encoded.size() < SmallString::kMaxSize)) {
```
Comment (romange) [inline src/core/compact_object.cc:794]: probably `encoded.size() <= kMaxSize`

File: src/core/small_string.h (hunk 1 above)
Comment (romange) [reply src/core/small_string.h:59]: no need to change.

File: src/core/small_string.h [hunk 4]
```diff
@@ -55,7 +57,7 @@ class SmallString {
...
-  uint16_t size_;       // uint16_t - total size (including prefix)
+  uint8_t size_;        // uint8_t - total size (including prefix)
```
Comment (romange) [inline src/core/small_string.h:60]: i suggest keeping size_ as `uint16_t` - i have no desire to improve the efficiency of this optimization

File: src/core/small_string.h [hunk 5]
```diff
@@ -10,16 +10,18 @@
...
 namespace dfly {
 
-// blob strings of upto ~64KB. Small sizes are probably predominant
+// blob strings of upto ~256B. Small sizes are probably predominant
 // for in-memory workloads, especially for keys.
 // Please note that this class does not have automatic constructors and destructors, therefore
 // it requires explicit management.
 class SmallString {
-  static constexpr unsigned kPrefLen = 10;
+  static constexpr unsigned kPrefLen = 11;
```
Comment (romange) [inline src/core/small_string.h:18]: let's keep it 10 as it was before

**Review by romange — APPROVED** [review]

**Pull request merged**